core/src/main/scala/org/broadinstitute/dsde/rawls/fastpass/FastPassService.scala

```diff
@@ -0,0 +1,417 @@
+                                            childWorkspace: Workspace
+  ): ReadWriteAction[Unit] = {
+    if (!config.enabled) {
+      logger.debug(s"FastPass is disabled. Will not grant FastPass access to ${parentWorkspace.toWorkspaceName}")
+      return DBIO.successful()
+    }
+
+    DBIO.from(quotaAvailableForClonedWorkspaceFastPassGrants(parentWorkspace, childWorkspace)).flatMap {
+      quotaAvailable =>
+        if (quotaAvailable) {
+          logger.info(
+            s"Adding FastPass access for ${ctx.userInfo.userEmail} in workspace being cloned ${parentWorkspace.toWorkspaceName}"
+          )
+          val expirationDate = DateTime.now(DateTimeZone.UTC).plus(config.grantPeriod.toMillis)
+          for {
+            petEmail <- DBIO.from(samDAO.getUserPetServiceAccount(ctx, childWorkspace.googleProjectId))
+            userAndPet = UserAndPetEmails(WorkbenchEmail(ctx.userInfo.userEmail.value), petEmail)
+            _ <- setupBucketRoles(parentWorkspace, Set(SamWorkspaceRoles.reader), userAndPet, expirationDate)
+            _ <- DBIO
+              .from(openTelemetry.incrementCounter("fastpass-granted-user", tags = openTelemetryTags).unsafeToFuture())
+          } yield ()
+        } else {
+          logger.info(
+            s"Not enough IAM Policy Role Binding quota available to add FastPass access for ${ctx.userInfo.userEmail.value} in parent workspace ${parentWorkspace.toWorkspaceName}"
+          )
+          DBIO.successful()
+        }
+    }
+  }
+
+  def deleteFastPassGrantsForWorkspace(workspace: Workspace): ReadWriteAction[Unit] = {
+    logger.info(
+      s"Removing FastPass grants in workspace ${workspace.toWorkspaceName}"
+    )
+    for {
+      fastPassGrants <- dataAccess.fastPassGrantQuery.findFastPassGrantsForWorkspace(workspace.workspaceIdAsUUID)
+      _ <- removeFastPassGrantsInWorkspaceProject(fastPassGrants, workspace.googleProjectId)
+    } yield ()
+  }
+
+  private def removeFastPassGrantsInWorkspaceProject(fastPassGrants: Seq[FastPassGrant],
+                                                     googleProjectId: GoogleProjectId
+  ): ReadWriteAction[Unit] = {
+    val grantsByUserAndResource =
+      fastPassGrants.groupBy(g => (g.accountEmail, g.accountType, g.resourceType, g.resourceName))
+    val iamUpdates = grantsByUserAndResource.toSeq.map { grouped =>
+      val ((accountEmail, accountType, resourceType, resourceName), grants) = grouped
+      val organizationRoles = grants.map(_.organizationRole).toSet
+      val workbenchEmail = WorkbenchEmail(accountEmail.value)
+      val googleProject = GoogleProject(googleProjectId.value)
+      () =>
+        removeFastPassGrants(resourceType, resourceName, workbenchEmail, accountType, organizationRoles, googleProject)
+    }
+    for {
+      // this `flatMap` is necessary to run the IAM Updates in sequence, as to not sent conflicting policy updates to GCP
+      _ <- DBIO.from(iamUpdates.foldLeft(Future.successful())((a, b) => a.flatMap(_ => b())))
```

Comment:
perhaps a simplification for a later date: seems like this could be done with 2 total calls to google, 1 to update project iam and another to update bucket iam. Then you don't need to worry about serializing correctly